module/connection/ServerDialog.kt

```diff
@@ -78,48 +91,91 @@ object ServerDialog {
             }
         }
 
-        override fun trySubmit() {
+        override fun submit() {
             when (server) {
-                TYPEDB -> StudioState.client.tryConnectToTypeDBAsync(address) { StudioState.client.connectServerDialog.close() }
-                TYPEDB_CLUSTER -> when {
-                    caCertificate.isBlank() -> StudioState.client.tryConnectToTypeDBClusterAsync(
-                        address, username, password, tlsEnabled
-                    ) { StudioState.client.connectServerDialog.close() }
-                    else -> StudioState.client.tryConnectToTypeDBClusterAsync(
-                        address, username, password, caCertificate
-                    ) { StudioState.client.connectServerDialog.close() }
+                TYPEDB ->  {
+                    StudioState.client.tryConnectToTypeDBAsync(coreAddress) {
+                        StudioState.client.connectServerDialog.close()
+                    }
+                }
+                TYPEDB_CLUSTER -> {
+                    when {
+                        caCertificate.isBlank() -> StudioState.client.tryConnectToTypeDBClusterAsync(
+                            clusterAddresses.toSet(), username, password, tlsEnabled
+                        ) {
+                            StudioState.client.connectServerDialog.close()
+                        }
+                        else -> StudioState.client.tryConnectToTypeDBClusterAsync(
+                            clusterAddresses.toSet(), username, password, caCertificate
+                        ) {
+                            StudioState.client.connectServerDialog.close()
+                        }
+                    }
                 }
             }
             appData.server = server
-            appData.address = address
+            appData.coreAddress = coreAddress
+            appData.clusterAddresses = clusterAddresses
             appData.username = username
             appData.tlsEnabled = tlsEnabled
             appData.caCertificate = caCertificate
         }
     }
 
+    private object CreateAddressForm : Form.State() {
+
+        var value: String by mutableStateOf("")
+
+        override fun cancel() {
+            StudioState.client.manageAddressesDialog.close()
+        }
+
+        override fun isValid(): Boolean {
+            return value.isNotBlank() && validAddress() && !state.clusterAddresses.contains(value)
+        }
+
+        override fun submit() {
+            assert(isValid())
+            state.clusterAddresses.add(value)
+            value = ""
+        }
+
+        private fun validAddress(): Boolean {
+            val addressParts = value.split(":")
+            if (addressParts.size == 2 && addressParts[1].toIntOrNull() != null) {
+                val port = addressParts[1].toInt()
+                if (port in 1024..65535) {
+                    return true
+                }
+            }
+            return false
+        }
```

Comment:
We now validate that the address input is of the correct form. We only verify that the address if of the form address:port and that the port is valid as hostnames can take broadly any form. 

We verify that the port is valid by ensuring that it is an integer and that it complies with IANA's port layout (ports 0-1023 are reserved system ports that typically require superuser access to bind to, and ports go up to 65535) https://www.rfc-editor.org/rfc/rfc1340.